Source/Core/Core/NetworkCaptureLogger.cpp

```diff
@@ -0,0 +1,378 @@
+  const std::string filename =
+      File::GetUserPath(D_DUMPSSL_IDX) + SConfig::GetInstance().GetGameID() + buffer;
+  m_file = std::make_unique<PCAP>(new File::IOFile(filename, "wb"), 1);
+}
+
+PCAPSSLCaptureLogger::ErrorState PCAPSSLCaptureLogger::SaveState() const
+{
+  return {
+      errno,
+#ifdef _WIN32
+      WSAGetLastError(),
+#endif
+  };
+}
+
+void PCAPSSLCaptureLogger::RestoreState(const PCAPSSLCaptureLogger::ErrorState& state) const
+{
+  errno = state.error;
+#ifdef _WIN32
+  WSASetLastError(state.wsa_error);
+#endif
+}
+
+void PCAPSSLCaptureLogger::Log(LogType log_type, bool is_sender, const void* data,
+                               std::size_t length, s32 socket)
+{
+  const auto state = SaveState();
+  if (log_type == LogType::UDP)
+    LogUDP(is_sender, data, length, socket);
+  else
+    LogTCP(is_sender, data, length, socket);
+  RestoreState(state);
+}
+
+void PCAPSSLCaptureLogger::LogSSLRead(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::SSL, false, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogSSLWrite(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::SSL, true, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogTCPRead(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::TCP, false, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogTCPWrite(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::TCP, true, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogUDPRead(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::UDP, false, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogUDPWrite(const void* data, std::size_t length, s32 socket)
+{
+  Log(LogType::UDP, true, data, length, socket);
+}
+
+void PCAPSSLCaptureLogger::LogRead(const void* data, std::size_t length, s32 socket)
+{
+  int socket_type;
+  socklen_t option_length = sizeof(int);
+
+  const auto state = SaveState();
+  if (getsockopt(socket, SOL_SOCKET, SO_TYPE, reinterpret_cast<char*>(&socket_type),
+                 &option_length) == 0)
+  {
+    if (socket_type == SOCK_STREAM)
+      LogTCPRead(data, length, socket);
+    else if (socket_type == SOCK_DGRAM)
+      LogUDPRead(data, length, socket);
+  }
+  RestoreState(state);
+}
+
+void PCAPSSLCaptureLogger::LogWrite(const void* data, std::size_t length, s32 socket)
+{
+  int socket_type;
+  socklen_t option_length = sizeof(int);
+
+  const auto state = SaveState();
+  if (getsockopt(socket, SOL_SOCKET, SO_TYPE, reinterpret_cast<char*>(&socket_type),
+                 &option_length) == 0)
+  {
+    if (socket_type == SOCK_STREAM)
+      LogTCPWrite(data, length, socket);
+    else if (socket_type == SOCK_DGRAM)
+      LogUDPWrite(data, length, socket);
+  }
+  RestoreState(state);
+}
+
+std::vector<u8> PCAPSSLCaptureLogger::GetEthernetHeader(bool is_sender)
+{
+  constexpr std::size_t ETHERNET_HEADER_SIZE = 14;
+  std::vector<u8> ethernet_header(ETHERNET_HEADER_SIZE, 0);
```

Comment:
`, 0` can be omitted. By default vectors zero out their contents on construction.